### PR TITLE
Align RT-1.26 Basic Static Route test with README and implement missing tests

### PR DESCRIPTION
**Pull request by vjarial-web**
**Objective:** Validate foundational routing capabilities for static routes on a Device Under Test (DUT), ensuring robust support for IPv4/IPv6 routing operations, multi-path topologies, configuration attributes, and negative edge-cases as rigorously outlined in RT-1.26.

**Interfaces Used:** gNMI, IPv4, IPv6, IS-IS, Traffic Generator (OTG)

**Topology Note:** Requires a standard 4-link ATE-DUT-ATE topology setup (`featureprofiles/topologies/atedut_4.testbed`).

**Expected Outcome Summary:**

*   **Run 1 (Static Route Verification & Attributes):** Validate foundational routing logic by deploying Equal-Cost Multi-Path (ECMP) scenarios to split packet transmissions equally across connected endpoints. Adjust OpenConfig structures to prioritize identical prefixes via deterministic metric manipulation and administrative distance (preference) adjustments. Append arbitrary route tags and verify their retention sequentially across gNMI state channels.
*   **Run 2 (Advanced Next-Hops & Negative Testing):** Verify Cross-Address Family (XAF) bridging by directing IPv4 destination routes through IPv6 next-hops, and vice versa. Intentionally blackhole flows utilizing local `DROP` next-hops( test intentionally sending streams of traffic to the next-hop, then Openconfig DROP the traffic or DUMP it to see 100% drop). Explicitly disable recursive sequence lookups via IS-IS endpoints to confirm packets die gracefully instead of traversing backdoors. Ensure device stability during dynamic scaling of active next-hops and spontaneous Direct Interface IP deletions. Verify Longest Prefix Match (LPM) superiority through overlapping IPv4 subsets and guarantee OS elasticity by deliberately provoking and capturing route resolution recursion loops.

**Success Metrics:**

*   **Routing Accuracy:** The DUT seamlessly executes targeted forwarding protocols dictated by dynamic parameters (ECMP, Metric, Preference) and successfully routes Cross-Address Family traffic without data plane interruption.
*   **Control Plane Resilience:** The DUT infrastructure successfully processes negative and corner-case topology conditions. The OS does not hang, crash, or incorrectly forward traffic when facing routing resolution loops, explicitly dropped next-hops, or sudden recursive path closures.

### CODE REVIEW

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go [hunk 1]
```diff
@@ -1447,3 +1322,349 @@ func (td *testData) advertiseRoutesWithISIS(t *testing.T) {
...
+
+		portCountersV4 := egressTrackingCounters(t, td.ate, v4Flow)
+		p1CounterV4, ok4_1 := portCountersV4[port1Tag]
+		p2CounterV4, ok4_2 := portCountersV4[port2Tag]
+
+		if ok4_1 && ok4_2 {
+			if got, want := p1CounterV4*100/(p1CounterV4+p2CounterV4), uint64(50); got < want-ecmpTolerance || got > want+ecmpTolerance {
+				t.Errorf("ECMP IPv4 load balance error for port1, got: %v, want: %v", got, want)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (vjarial-web) [reply feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go:1461]: Verifying Equal-Cost Multi-Path (ECMP) routing. 1. Calculate the total first  2. If the total is 0, softly fail the test with a clear, readable error message. 3. Otherwise, it is safe to divide! Calculate the percentage and test if it's roughly 50%.   We are checking if the traffic was split 50/50 between port1 and port2.   Fixing the initial 0 traffic bug.  Ipv4

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go [hunk 2]
```diff
@@ -1447,3 +1322,349 @@ func (td *testData) advertiseRoutesWithISIS(t *testing.T) {
...
+
+		portCountersV6 := egressTrackingCounters(t, td.ate, v6Flow)
+		p1CounterV6, ok6_1 := portCountersV6[port1Tag]
+		p2CounterV6, ok6_2 := portCountersV6[port2Tag]
+		if ok6_1 && ok6_2 {
+			if got, want := p1CounterV6*100/(p1CounterV6+p2CounterV6), uint64(50); got < want-ecmpTolerance || got > want+ecmpTolerance {
+				t.Errorf("ECMP IPv6 load balance error for port1, got: %v, want: %v", got, want)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (vjarial-web) [reply feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go:1472]: Verifying Equal-Cost Multi-Path (ECMP) routing. 1. Calculate the total first  2. If the total is 0, softly fail the test with a clear, readable error message. 3. Otherwise, it is safe to divide! Calculate the percentage and test if it's roughly 50%.   We are checking if the traffic was split 50/50 between port1 and port2.   Fixing the initial 0 traffic bug.  for IPv6

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go [hunk 3]
```diff
@@ -1447,3 +1322,349 @@ func (td *testData) advertiseRoutesWithISIS(t *testing.T) {
...
+
+	portCounters := egressTrackingCounters(t, td.ate, lpmFlowName)
+	p2Counter, ok := portCounters[port2Tag]
+	if !ok {
+		t.Logf("LPM tracking counters: %v", portCounters)
+	}
+	_, rxLPM := otgutils.GetFlowStats(t, td.ate.OTG(), lpmFlowName, 20*time.Second)
+	if got, want := float64(p2Counter)*100/float64(rxLPM), float64(100); got+lossTolerance < want {
+		t.Errorf("LPM IPv4 traffic on port2, got: %v%%, want: %v%%", got, want)
+	}
```
[the remark this answers is not in the document]
Comment (vjarial-web) [reply feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go:1630]: Fixing the testOverlappingPrefixesLPM , test configures two overlapping routes (10.0.0.0/8 pointing towards ATE port-1) and (10.1.1.0/24 pointing towards ATE port-2), Traffic generator is sending traffic for 10.1.1.1, so 100% traffic will go to longest prefix match, To avoid 0 traffic issue we are fixing the float64(p2Counter)*100/float64(rxLPM)  code fatil error.

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go [hunk 4]
```diff
@@ -1447,3 +1315,335 @@ func (td *testData) advertiseRoutesWithISIS(t *testing.T) {
...
+			"1": oc.UnionString(atePort2.IPv6),
+		},
+	}
+	if _, err := cfgplugins.NewStaticRouteCfg(b, v4Cfg, td.dut); err != nil {
+		t.Fatalf("Failed to configure IPv4 static route: %v", err)
+	}
+
+	// Step 4 - Push configuration to DUT.
+	b.Set(t, td.dut)
+	defer td.deleteStaticRoutes(t)
```
[the remark this answers is not in the document]
Comment (vjarial-web) [reply feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go]: Made changes for successful test registering cleanup operation.

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go [hunk 5]
```diff
@@ -1447,3 +1315,335 @@ func (td *testData) advertiseRoutesWithISIS(t *testing.T) {
...
+func (td *testData) testCrossAddressFamilyNextHops(t *testing.T) {
+	if deviations.IPv6StaticRouteWithIPv4NextHopUnsupported(td.dut) || deviations.IPv4StaticRouteWithIPv6NextHopUnsupported(td.dut) {
+		t.Skip("Skipping XAF route test. Deviations unsupported.")
+	}
+
+	// Step 1 - Delete the configuration using a gNMI Set DELETE on the specific `metric` and `preference` paths for `ipv4-route-b`, `ipv6-route-b`, `ipv4-route-a`, and `ipv6-route-a`.
+	cfgplugins.DeleteStaticRouteNextHopLeaves(t, td.dut, deviations.DefaultNetworkInstance(td.dut), td.staticIPv4.cidr(t), "0", "metric", "preference")
+	cfgplugins.DeleteStaticRouteNextHopLeaves(t, td.dut, deviations.DefaultNetworkInstance(td.dut), td.staticIPv4.cidr(t), "1", "metric", "preference")
+	cfgplugins.DeleteStaticRouteNextHopLeaves(t, td.dut, deviations.DefaultNetworkInstance(td.dut), td.staticIPv6.cidr(t), "0", "metric", "preference")
+	cfgplugins.DeleteStaticRouteNextHopLeaves(t, td.dut, deviations.DefaultNetworkInstance(td.dut), td.staticIPv6.cidr(t), "1", "metric", "preference")
```
[the remark this answers is not in the document]
Comment (vjarial-web) [reply feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go]: Removed the redundant DeleteStaticRouteNextHopLeaves checks. The routes are guaranteed to be purged by the t.Cleanup() blocks in the preceding subtests, so we are starting with a clean slate here.

File: internal/cfgplugins/staticroute.go [hunk 6]
```diff
@@ -244,6 +311,30 @@ func ConfigureStaticRoute(t *testing.T, dut *ondatra.DUTDevice, batch *gnmi.SetB
...
+	} else {
+		// Validate both the routes i.e. ipv4-route-[a|b] are configured and reported
+		// correctly
+		gotStatic := gnmi.Get(t, dut, sp.Static(prefix).State())
+		t.Logf("Static route %s: got: %v, want: %v", prefix, len(gotStatic.NextHop), len(sV4.NextHops))
+		for index, nextHop := range gotStatic.NextHop {
+			if got, want := nextHop.GetNextHop(), sV4.NextHops[index]; got != want {
+				t.Errorf("Static route %s: got: %v, want: %v", prefix, got, want)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (vjarial-web) [reply internal/cfgplugins/staticroute.go:336]: Before : This causes the test to silently pass, even though some expected next hops are missing. Now : It explicitly check that the total count of retrieved next hops matches the expected count.  raise error

File: internal/cfgplugins/staticroute.go [hunk 7]
```diff
@@ -216,16 +232,67 @@ func NewStaticVRFRoute(t *testing.T, batch *gnmi.SetBatch, cfg *StaticVRFRouteCf
...
+	b.Set(t, dut)
+}
+
+// ValidateStaticRouteNextHopIndex validates the next hop indexes and their addresses.
+func ValidateStaticRouteNextHopIndex(t *testing.T, dut *ondatra.DUTDevice, netInst string, prefix string, expectedNh map[string]string) {
+	t.Helper()
+	ni := normalizeNIName(netInst, dut)
+	sp := gnmi.OC().NetworkInstance(ni).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_STATIC, deviations.StaticProtocolName(dut))
+
+	gotStatic := gnmi.Get(t, dut, sp.Static(prefix).State())
```
[the remark this answers is not in the document]
Comment (vjarial-web) [reply internal/cfgplugins/staticroute.go:280]: gnmi.Await  is added so it will wait for device status for 30 sec, then run gnmi.get.  This make test stable and reliable

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go [hunk 8]
```diff
@@ -1447,3 +1315,335 @@ func (td *testData) advertiseRoutesWithISIS(t *testing.T) {
...
+
+	// Step 3: Push configuration to DUT.
+	b.Set(t, td.dut)
+	defer func() {
+		bDel := &gnmi.SetBatch{}
+		sp := gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(td.dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_STATIC, deviations.StaticProtocolName(td.dut))
+		gnmi.BatchDelete(bDel, sp.Static(prefixA).Config())
+		gnmi.BatchDelete(bDel, sp.Static(prefixB).Config())
+		bDel.Set(t, td.dut)
+	}()
```
[the remark this answers is not in the document]
Comment (vjarial-web) [reply feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go]: Before we are first,  directly pushing the configuration on device.  Now, if the b.Set(t, td.dut) step fails and immediately aborts the test, the t.Cleanup() routine will still execute.

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go [hunk 9]
```diff
@@ -1447,3 +1315,335 @@ func (td *testData) advertiseRoutesWithISIS(t *testing.T) {
...
+
+	// Step 2: Push configuration to DUT.
+	b.Set(t, td.dut)
+	defer func() {
+		bDel := &gnmi.SetBatch{}
+		sp := gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(td.dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_STATIC, deviations.StaticProtocolName(td.dut))
+		gnmi.BatchDelete(bDel, sp.Static(prefix8).Config())
+		gnmi.BatchDelete(bDel, sp.Static(prefix24).Config())
+		bDel.Set(t, td.dut)
+	}()
```
[the remark this answers is not in the document]
Comment (vjarial-web) [reply feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go]: Before we are first, directly pushing the configuration on device. Now, if the b.Set(t, td.dut) step fails and immediately aborts the test, the t.Cleanup() routine will still execute.

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go [hunk 10]
```diff
@@ -1447,3 +1315,335 @@ func (td *testData) advertiseRoutesWithISIS(t *testing.T) {
...
+	gnmi.Delete(t, td.dut, gnmi.OC().Interface(port2).Subinterface(0).Ipv4().Address(dutPort2.IPv4).Config())
+
+	defer func() {
+		ipConf := &oc.Interface_Subinterface_Ipv4_Address{
+			Ip:           ygot.String(dutPort2.IPv4),
+			PrefixLength: ygot.Uint8(uint8(dutPort2.IPv4Len)),
+		}
+		gnmi.Replace(t, td.dut, gnmi.OC().Interface(port2).Subinterface(0).Ipv4().Address(dutPort2.IPv4).Config(), ipConf)
+		td.deleteStaticRoutes(t)
+	}()
```
[the remark this answers is not in the document]
Comment (vjarial-web) [reply feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go]: I previous  code if gnmi.Delete  failed defer block will never execute. 
 Now [By registering the t.Cleanup() routine first, the recovery instructions are safely queued up. Now, if the gnmi.Delete step fails and immediately halts the test, the cleanup code will still execute to restore the interface IP (gnmi.Replace) and remove the static routes (td.deleteStaticRoutes(t)), ensuring the device is always reset to a clean state.]

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go [hunk 11]
```diff
@@ -243,55 +255,64 @@ func TestStaticRouteAddRemove(t *testing.T) {
...
-	}
-	if _, err := cfgplugins.NewStaticRouteCfg(b, sV4, dut); err != nil {
-		t.Fatalf("Failed to configure IPv4 static route: %v", err)
-	}
-	b.Set(t, dut)
-	validateStaticRoute(t, dut, prefix.cidr(t), sV4)
+	t.Run("RT-1.26.8 - Validate Dynamic Add and Remove of Next-Hops", func(t *testing.T) {
+		// Step 1 - Configure one IPv4 static route with next-hops set to the IPv4
+		// address of ATE port-2 (index 0) and port-3 (index 1).
+		prefix := ipAddr{address: v4Route, prefix: v4RoutePrefix}
```
[the remark this answers is not in the document]
Comment (vjarial-web) [reply feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go:253]: Cleaning the device state after test run.

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go [hunk 12]
```diff
@@ -1447,3 +1314,355 @@ func (td *testData) advertiseRoutesWithISIS(t *testing.T) {
...
+
+	// FIX: Add gnmi Wait block to verify the IP address has actually been removed from State!
+	t.Log("Awaiting state convergence for Interface IP deletion...")
+	gnmi.Watch(t, td.dut, gnmi.OC().Interface(port2).Subinterface(0).Ipv4().Address(dutPort2.IPv4).State(), 30*time.Second, func(val *ygnmi.Value[*oc.Interface_Subinterface_Ipv4_Address]) bool {
+		return !val.IsPresent()
+	}).Await(t)
+
+	// Step 3: Send Traffic
+	td.ate.OTG().StartTraffic(t)
+	time.Sleep(trafficDuration)
```
Comment (navaneethyv) [inline feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go:1529]: @AmrNJ  what is the recommended way here and why is it not caught by Gemini ?

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go [hunk 13]
```diff
@@ -276,22 +272,32 @@ func TestStaticRouteAddRemove(t *testing.T) {
...
 	}
 	if _, err := cfgplugins.NewStaticRouteCfg(b, sV4, dut); err != nil {
 		t.Fatalf("Failed to configure IPv4 static route: %v", err)
 	}
+	// Note: We need to explicitly delete the next-hops index 0 and 3
+	// using the new helper or manually
+	cfgplugins.DeleteStaticRouteNextHopLeaves(t, dut, deviations.DefaultNetworkInstance(dut), prefix.cidr(t), "0", "metric", "preference", "recurse") // Just to clear config
+
+	// Actually we should just set the batch for the whole route, but the route replacement doesn't delete old leaves in OC unless requested.
+	// Oh wait, NewStaticRouteCfg just sets the leaves that we give, it doesn't delete the other leaves unless we do gnmi.Update on the whole StaticRoute list which isn't done explicitly. Wait, actually NewStaticRouteCfg does gnmi.BatchReplace on the specific prefix level configuration. So replacing the prefix configuration drops old next-hops.
```
Comment (navaneethyv) [inline feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go:296]: This seems like a straight out AI comment, can we double check these on the whys and add meaningful shorter insights that doesn't add cognitive load to the users ?

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go (hunk 12 above)
Comment (AmrNJ) [reply feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go:1529]: We allow sleep between Start and Stop traffic. Gemini review only comments on usage of sleep wherever telemetry paths are polled.

Comment (navaneethyv) [reply feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go:1529]: Thx.

**Comment by navaneethyv** [comment]
Please have a description of the changes in the PR instead of generic description as this will be reviewed across.

File: feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go (hunk 13 above)
Comment (vjarial-web) [reply feature/staticroute/otg_tests/basic_static_route_support_test/basic_static_route_support_test.go]: This issue is fixed relating to AI comments.

**Review by rohit-rp — APPROVED** [review]
